File: src/apps/quat.cpp
// ========================================================================== //
// Copyright (c) 2014-2019 The University of Texas at Austin.                 //
// All rights reserved.                                                       //
//                                                                            //
// Licensed under the Apache License, Version 2.0 (the "License");            //
// you may not use this file except in compliance with the License.           //
// A copy of the License is included with this software in the file LICENSE.  //
// If your copy does not contain the License, you may obtain a copy of the    //
// License at:                                                                //
//                                                                            //
//     https://www.apache.org/licenses/LICENSE-2.0                            //
//                                                                            //
// Unless required by applicable law or agreed to in writing, software        //
// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT  //
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.           //
// See the License for the specific language governing permissions and        //
// limitations under the License.                                             //
//                                                                            //
// ========================================================================== //

#if 1
void ksdflksaf() {}
#else
#include <iostream>
#include "quat.h"
#include <math.h>

namespace gxy
{
#define TRACKBALLSIZE 2.0

static float
project_to_sphere(float r, float x, float y)
{
    float d, t, z;

    d = sqrt(x*x + y*y);
    if (d < r * 0.70710678118654752440)
        z = sqrt(r*r - d*d);
    else
		{
        t = r / 1.41421356237309504880;
        z = t*t / d;
    }
    return z;
}


void axis_to_quat(vec3f a, float theta, vec4f& q)
{
		std::cerr << "axis " << a.x << " " << a.y << " " << a.z << " theta " << theta << "\n";

		// set(q, cos(theta/2), a.x*sin(theta/2), a.y*sin(theta/2), a.z*sin(theta/2));

		float t = a.x*sin(theta/2);
		float x = a.y*sin(theta/2);
		float y = a.z*sin(theta/2);
		float z = cos(theta/2);

#if 0
		float d = sqrt(x*x + y*y + z*z);
		if (d > 0)
		{
			d = 1.0 / d;
			x *= d;
			y *= d;
			z *= d;
		}
#endif
			
		set(q, t, x, y, z);
		
		std::cerr << "q " << q.x << " " << q.y << " " << q.z << " " << q.w << "\n";
}

void multiply_quats(vec4f q1, vec4f q2, vec4f& dest)
{
	dest.x = q1.x*q2.x - q1.y*q2.y - q1.z*q2.z - q1.w*q2.w;
	dest.y = q1.x*q2.y + q1.y*q2.x + q1.z*q2.w - q1.w*q2.z;
	dest.z = q1.x*q2.z - q1.y*q2.w + q1.z*q2.x + q1.w*q2.y;
	dest.w = q1.x*q2.w + q1.y*q2.z - q1.z*q2.y + q1.w*q2.x;
}

void add_quats(vec4f q1, vec4f q2, vec4f& dest)
{
    vec4f t1, t2, t3;
    vec4f tf;

    t1 = q1;
    scale(q2.w, t1);

    t2 = q2;
    scale(q1.w, t2);

    cross(q2,q1,t3);
    add(t1,t2,tf);
    add(t3,tf,tf);
    tf.w = q1.w * q2.w - dot(q1,q2);

    dest = tf;
		normalize(dest);
}

void
rotate_vector_by_quat(vec3f v, vec4f q, vec3f& vq)
{
	vec3f u = {q.y, q.z, q.w};
	float s = q.x;

	float dot_uv = dot(u, v);
	float dot_uu = dot(u, u);
	float ss     = s*s;

	vec3f cross_uv;
	cross(u, v, cross_uv);

	float a = 2.0 * dot(u, v);
	float b = ss - dot(u, u);
	float c = 2 * s;

    vq.x = (a * u.x) + (b * v.x) + (c * cross_uv.x);
    vq.y = (a * u.y) + (b * v.y) + (c * cross_uv.y);
    vq.z = (a * u.z) + (b * v.z) + (c * cross_uv.z);

		float d = 1.0 / sqrt(vq.x*vq.x + vq.y*vq.y + vq.z*vq.z);
    vq.x *= d;
    vq.y *= d;
    vq.z *= d;

}

void
trackball(vec4f& q, float p1x, float p1y, float p2x, float p2y)
{
    vec3f a; /* Axis of rotation */
    float phi;  /* how much to rotate about axis */
    vec3f p1, p2, d;
    float t;

    if (p1x == p2x && p1y == p2y)
		{
        zero(q);
        return;
    }

    /*
     * First, figure out z-coordinates for projection of P1 and P2 to
     * deformed sphere
     */
    p1 = vec3f(p1x, p1y, project_to_sphere(TRACKBALLSIZE,p1x,p1y));
    p2 = vec3f(p2x, p2y, project_to_sphere(TRACKBALLSIZE,p2x,p2y));

std::cerr << "===============================\n";
std::cerr << "p1 " << p1.x << " " << p1.y << " " << p1.z << "\n";
std::cerr << "p2 " << p2.x << " " << p2.y << " " << p2.z << "\n";

    /*
     *  Now, we want the cross product of P1 and P2
     */
    cross(p2,p1,a);
    normalize(a);

std::cerr << "A: " << a.x << " " << a.y << " " << a.z << "\n";

    /*
     *  Figure out how much to rotate around that axis.
     */
    sub(p1,p2,d);
    t = len(d) / (2.0*TRACKBALLSIZE);

    /*
     * Avoid problems with out-of-control values...
     */
    if (t > 1.0) t = 1.0;
    if (t < -1.0) t = -1.0;
    phi = 2.0 * asin(t);

    axis_to_quat(a,phi,q);
}

}
#endif

File: src/cinema/quat.py
## ========================================================================== ##
## Copyright (c) 2014-2019 The University of Texas at Austin.                 ##
## All rights reserved.                                                       ##
##                                                                            ##
## Licensed under the Apache License, Version 2.0 (the "License");            ##
## you may not use this file except in compliance with the License.           ##
## A copy of the License is included with this software in the file LICENSE.  ##
## If your copy does not contain the License, you may obtain a copy of the    ##
## License at:                                                                ##
##                                                                            ##
##     https://www.apache.org/licenses/LICENSE-2.0                            ##
##                                                                            ##
## Unless required by applicable law or agreed to in writing, software        ##
## distributed under the License is distributed on an "AS IS" BASIS, WITHOUT  ##
## WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.           ##
## See the License for the specific language governing permissions and        ##
## limitations under the License.                                             ##
##                                                                            ##
## ========================================================================== ##

from math import sqrt, sin, cos, asin

TRACKBALLSIZE = 0.8

def cross(a, b):
	return [a[1]*b[2] - a[2]*b[1], a[2]*b[0] - a[0]*b[2], a[0]*b[1] - a[1]*b[0]]

def length(vec):
  sum = 0
  for v in vec:
    sum = sum + (v*v)
  return sqrt(sum)

def normalize(vec):
  l = length(vec)
  return [i/l for i in vec]

def scale(s, vec):
  return [s*i for i in vec]

def addv(a, b):
  c = [i[0] + i[1] for i in zip(a, b)]
  return c

def subv(a, b):
  c = [i[0] - i[1] for i in zip(a, b)]
  return c

def dot(a, b):
  sum = 0
  for i in zip(a[:3], b[:3]):
    sum = i[0]*i[1]
  return sum

def project_to_sphere(r, x, y):
    d = sqrt(x*x + y*y)
    if d < (r * 0.70710678118654752440):
        z = sqrt(r*r - d*d)
    else:
        t = r / 1.41421356237309504880
        z = t*t / d
    return z

def axis_to_quat(a, phi):
    a = normalize(a)
    if phi == 0:
			return [0, 0, 0, 1]
    else:
      a = scale(sin(phi/2.0), a)
      return [a[0], a[1], a[2], cos(phi/2.0)]

def add_quats(q1, q2):
    t1 = scale(q2[3], q1)
    t2 = scale(q1[3], q2)
    t3 = cross(q1[:3], q2[:3])
    tf = addv(t1,t2)
    tf = addv(t3,tf)
    tf = tf + [q1[3] * q2[3] - dot(q1,q2)]
    return normalize(tf)

def rotate_vector_by_quat(v, q):
     num = q[0] * 2.0
     num2 = q[1] * 2.0
     num3 = q[2] * 2.0
     num4 = q[0] * num
     num5 = q[1] * num2
     num6 = q[2] * num3
     num7 = q[0] * num2
     num8 = q[0] * num3
     num9 = q[1] * num3
     num10 = q[3] * num
     num11 = q[3] * num2
     num12 = q[3] * num3
     return [(1.0 - (num5 + num6)) * v[0] + (num7 - num12) * v[1] + (num8 + num11) * v[2],
             (num7 + num12) * v[0] + (1.0 - (num4 + num6)) * v[1] + (num9 - num10) * v[2],
             (num8 - num11) * v[0] + (num9 + num10) * v[1] + (1.0 - (num4 + num5)) * v[2]]

def trackball(p1x, p1y, p2x, p2y):
    if p1x == p2x and p1y == p2y:
        return [0.0, 0.0, 0.0, 0.0]

    p1 = [p1x, p1y, project_to_sphere(TRACKBALLSIZE,p1x,p1y)]
    p2 = [p2x, p2y, project_to_sphere(TRACKBALLSIZE,p2x,p2y)]

    a = normalize(cross(p2,p1))
    d = sub(p1,p2)

    t = length(d) / (2.0*TRACKBALLSIZE)
    if t > 1.0:
       t = 1.0
    if t < -1.0:
       t = -1.0
    phi = 2.0 * asin(t)

    return axis_to_quat(a,phi)
